### PR TITLE
Msys2 ci

### PR DESCRIPTION
**Pull request by Kreijstal**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**
Hello, this is my attempt to add CI, I was not sure if I should do it under CI or build, please check that I still can't build with meson on msys2 :) #23691

### CODE REVIEW

**Comment by Kreijstal** [comment]
You can see the workflow execution here https://github.com/Kreijstal/radare2/actions/runs/12072993748/job/33668184164

**Comment by trufae** [comment]
try this :

```
0$ git diff
diff --git a/libr/include/r_util/r_w32.h b/libr/include/r_util/r_w32.h
index a16be90a47..d9928c4368 100644
--- a/libr/include/r_util/r_w32.h
+++ b/libr/include/r_util/r_w32.h
@@ -5,7 +5,7 @@
 extern "C" {
 #endif

-#if R2__WINDOWS__
+#if R2__WINDOWS__ || __MINGW32__
 #include <windows.h> // CreateToolhelp32Snapshot
 #include <tlhelp32.h> // CreateToolhelp32Snapshot
 #include <psapi.h> // GetModuleFileNameEx, GetProcessImageFileName
0$
```

File: libr/util/sandbox.c [hunk 1]
```diff
@@ -15,6 +15,10 @@
 #include <priv.h>
 #endif
 
+#ifdef __MINGW32__
+#include <unistd.h>
```
Comment (trufae) [inline libr/util/sandbox.c:19]: maybe worth adding it in r_types.h, so other files can be affected

Comment (Kreijstal) [reply libr/util/sandbox.c:19]: the error comes again after installing in r_types.h

Comment (Kreijstal) [reply libr/util/sandbox.c]: Done

**Comment by Kreijstal** [comment]
> try this :
> 
> ```
> 0$ git diff
> diff --git a/libr/include/r_util/r_w32.h b/libr/include/r_util/r_w32.h
> index a16be90a47..d9928c4368 100644
> --- a/libr/include/r_util/r_w32.h
> +++ b/libr/include/r_util/r_w32.h
> @@ -5,7 +5,7 @@
>  extern "C" {
>  #endif
> 
> -#if R2__WINDOWS__
> +#if R2__WINDOWS__ || __MINGW32__
>  #include <windows.h> // CreateToolhelp32Snapshot
>  #include <tlhelp32.h> // CreateToolhelp32Snapshot
>  #include <psapi.h> // GetModuleFileNameEx, GetProcessImageFileName
> 0$
> ```

Thank you but, it seems didn't work.

**Comment by trufae** [comment]
the implmenetaiton is in here:
```
0$ git diff
diff --git a/libr/util/w32.c b/libr/util/w32.c
index 974ef2f16d..8d73090ba4 100644
--- a/libr/util/w32.c
+++ b/libr/util/w32.c
@@ -3,7 +3,7 @@
 #include <r_userconf.h>
 #include <r_util.h>

-#if R2__WINDOWS__
+#if R2__WINDOWS__ || __MINGW32__
 #include <windows.h>

 #if 1
0$
```

**Comment by Kreijstal** [comment]
also for some reason tcc is failing, did I break it?

**Comment by Kreijstal** [comment]
> the implmenetaiton is in here:
> 
> ```
> 0$ git diff
> diff --git a/libr/util/w32.c b/libr/util/w32.c
> index 974ef2f16d..8d73090ba4 100644
> --- a/libr/util/w32.c
> +++ b/libr/util/w32.c
> @@ -3,7 +3,7 @@
>  #include <r_userconf.h>
>  #include <r_util.h>
> 
> -#if R2__WINDOWS__
> +#if R2__WINDOWS__ || __MINGW32__
>  #include <windows.h>
> 
>  #if 1
> 0$
> ```

it seems R2__WINDOWS__ is already true on mingw platform, so this is a no-op

**Comment by trufae** [comment]
so maybe this?

```
diff --git a/libr/include/r_th.h b/libr/include/r_th.h
index 518dc0e067..c505290320 100644
--- a/libr/include/r_th.h
+++ b/libr/include/r_th.h
@@ -8,6 +8,7 @@
 #include "r_types.h"
 #include "r_userconf.h"
 #include <r_list.h>
+#include <r_util/r_w32.h>

 #include <stdio.h>
 #include <stdlib.h>
```

**Comment by trufae** [comment]
all those format warnings make me think that mingw uses fprintf from unix instead of the windows unix runtime. so the format must be %llx instead of %I64...

```

../libr/util/file.c:873:30: warning: unknown conversion type character 'I' in format [-Wformat=]
  873 |                 fprintf (fd, "0x%08"PFMT64x"  ", (ut64)i);
      |                              ^~~~~~~
```

can you try that fixing the ifdef in r_types.h:?
```
#define PFMT64x "I64x"
#define PFMT64x PRIx64
```

you can use #error statements in the include headers to see if those are included. my guess is that r_w32 is included twice, the first one without the R2__WINDOWS__ var defined, and the second with __MINGW__ around, so the include guard doesnt include the contents later.

**Comment by Kreijstal** [comment]
> #define PFMT64x "I64x"
> #define PFMT64x PRIx64

Oh wow that seemed to work, it's compiling beyond what it normally does.

**Comment by Kreijstal** [comment]
That's a great progress!
```
../libr/debug/p/debug_windbg.c:631:17: error: initialization of '_Bool (*)(RDebug *)' {aka '_Bool (*)(struct r_debug_t *)'} from incompatible pointer type 'int (*)(RDebug *)' {aka 'int (*)(struct r_debug_t *)'} [-Wincompatible-pointer-types]
  631 |         .stop = windbg_stop,
      |                 ^~~~~~~~~~~
```
now stuck on file [1070/1589] , I attempted to see the definition of RDebugPlugin found it but left very confused.

**Comment by trufae** [comment]
```
0$ git diff
diff --git a/libr/debug/p/debug_windbg.c b/libr/debug/p/debug_windbg.c
index de686d3e3a..b36fb8419c 100644
--- a/libr/debug/p/debug_windbg.c
+++ b/libr/debug/p/debug_windbg.c
@@ -116,7 +116,7 @@ static RDebugReasonType exception_to_reason(DWORD ExceptionCode) {
        }
 }

-static int windbg_stop(RDebug *dbg) {
+static bool windbg_stop(RDebug *dbg) {
        DbgEngContext *idbg = dbg->user;
        R_RETURN_VAL_IF_FAIL (idbg && idbg->initialized, 0);
        return SUCCEEDED (ITHISCALL (dbgCtrl, SetInterrupt, DEBUG_INTERRUPT_ACTIVE));
0$
```

that should fix this

**Comment by Kreijstal** [comment]
Finally! it compiled!

But I expect it to run a usage message
```
topkek@ayylmai UCRT64 ~/git/radare2
$ ./build/binr/radare2/radare2.exe

topkek@ayylmai UCRT64 ~/git/radare2
$ ./build/binr/radare2/radare2.exe -h

topkek@ayylmai UCRT64 ~/git/radare2
$ ./build/binr/radare2/radare2.exe --help

topkek@ayylmai UCRT64 ~/git/radare2
$ ./build/binr/r2pm/r2pm.exe

topkek@ayylmai UCRT64 ~/git/radare2
$ ./build/binr/r2r/r2r.exe

```
but it just empty, error level is `127`

```
topkek@ayylmai UCRT64 ~/git/radare2
# gdb ./build/binr/radare2/radare2.exe
GNU gdb (GDB) 15.2
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./build/binr/radare2/radare2.exe...
(gdb) b main
Breakpoint 1 at 0x1400017a3: file ../binr/radare2/radare2.c, line 102.
(gdb) r
Starting program: E:\msys64\home\topkek\git\radare2\build\binr\radare2\radare2.exe
[New Thread 10228.0x1654]
[New Thread 10228.0x9f0]
[New Thread 10228.0x1704]
[Thread 10228.0x1654 exited with code 3221225781]
[Thread 10228.0x29e8 exited with code 3221225781]
[Thread 10228.0x1704 exited with code 3221225781]
During startup program exited with code 0xc0000135.
(gdb)
```
It exits before it reaches main.

**Comment by trufae** [comment]
Thanks a lot! That’s amazing. Good work

**Pull request merged**